File: Leetcode/labuladong_template/java-template/src/main/java/leetcode/editor/cn/BinaryTreePreorderTraversal.java

package leetcode.editor.cn;

import java.sql.Array;
import java.util.*;
import leetcode.editor.common.*;

public class BinaryTreePreorderTraversal {

    //leetcode submit region begin(Prohibit modification and deletion)
    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     *     int val;
     *     TreeNode left;
     *     TreeNode right;
     *     TreeNode() {}
     *     TreeNode(int val) { this.val = val; }
     *     TreeNode(int val, TreeNode left, TreeNode right) {
     *         this.val = val;
     *         this.left = left;
     *         this.right = right;
     *     }
     * }
     */
    class Solution {
        List<Integer> res = new ArrayList<>();

        public List<Integer> preorderTraversal(TreeNode root) {
            traversal(root);
            return res;
        }

        void traversal(TreeNode root) {
            if (root == null) {
                return;
            }
            res.add(root.val);
            traversal(root.left);
            traversal(root.right);
        }
    }
    //leetcode submit region end(Prohibit modification and deletion)

    
    public static void main(String[] args) {

        Solution solution = new BinaryTreePreorderTraversal().new Solution();
        // put your test code here
        TreeNode root = TreeNode.createRoot(new Integer[]{1,2,3,4,5,null,8,null,null,6,7,9});

        List<Integer> result = solution.preorderTraversal(root);
        System.out.println(result);
    }
}

File: Leetcode/labuladong_template/js-template/leetcode/editor/cn/binary-tree-preorder-traversal.js
import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    const preOrderList = [];
    traverse(root, preOrderList);
    return preOrderList;
};

var traverse = function (root, preOrderList) {
    if (root === null) {
        return;
    }
    preOrderList.push(root.val);
    traverse(root.left, preOrderList);
    traverse(root.right, preOrderList);
}

//leetcode submit region end(Prohibit modification and deletion)

// your test code here
const root = TreeNode.createRoot([1,2,3,4,5,null,8,null,null,6,7,9])
console.log(preorderTraversal(root));
